python/cudf/cudf/core/dataframe.py

```diff
@@ -2184,6 +2184,84 @@ def melt(self, **kwargs):
 
         return melt(self, **kwargs)
 
+    def _typecast_before_merge(self, lhs, rhs, left_on, right_on, how):
+        def casting_rules(dtype_l, dtype_r, how):
+            cast_warn = "can't safely cast column {} from {} with type \
+                         {} to {}, upcasting to {}"
+            ctgry_err = "can't implicitly cast column {} to categories \
+                         from right during left join"
+
+            rtn = None
+            if pd.api.types.is_dtype_equal(dtype_l, dtype_r):
+                rtn = dtype_l
+            elif how == "left":
+
+                check_col = rhs._cols[rcol][rhs._cols[rcol].notna()]
+                if not (
+                    Series(check_col) == Series(check_col).astype(dtype_l)
+                ).all():
```

Comment:
Doing a boolmask gather, followed by a binary op, followed by an `all()` seems kinda expensive to see if we can safely typecast. Was this more performant than checking the max and min?

Alternatively, since this is to just ensure we're not overflowing, I think we can get away with not doing the boolmask gather and just doing a fillna=True, since we know we can always safely typecast nulls.

Reply:
@kkraus14 I did some hopefully careful thinking about this. 

Firstly a follow up to the boolmask question. I think I see what you're saying and believe this can be removed regardless of the rest of the approach, however don't we need the  `fillna()` before the binop? If we do the binop first I think we get back a column of bools so `fillna()` won't fill anything, but if we `fillna(0)` first then zero will always compare to zero and they'll be `True` for the `all()` which is what we want. That replaces the mask with a `fillna` op which could be helpful. 

Here are some thoughts on casting vs bounds checking to check for safe casting. In the min/max method, there's a few separate cases.

- The datatypes have the same `kind`. This has two subcases: (1) The left datatype is larger than the right and the cast is safe without proceeding further (the code should do this and currently doesn't, so let's ignore this case). (2) The left datatype is smaller than the right datatype, so we need to check the bounds, which I believe is a reduction each for the min and max and a binop each to compare each value of the column in question, and finally an `all()` to make sure all the values pass. (2 reductions, 2 binops involving `<` and `>`) 

- The left datatype is a float and the right is an int. Here there's a case where we could use `np.finfo` to figure out if the max and min of the ints are both within the range representable by a theoretical int with `np.finfo(the_float_datatype).nmant` bits. If yes then we're golden and this becomes the first case, but if not, we're (somewhat unfortunately) NOT necessarily sunk because the ints outside that range could be 'lucky' and fall along the real line in spots that the float still has exact coverage. As such I am not sure how to avoid explicitly attempting to cast each value and comparing it to the original. 
 
- The left datatype is an `int` and the right datatype is a `float`. In this case checking the min and max is a great start, but doesn't handle 'non int' floats like `5.5`, so in this case we need a binop to check against the min, one for the max, a third to mod the floats or similar and a fourth to compare the modded column with zero before the all. 

Ignoring the boolmask gather which should go, and the `all()` which we need no matter what, the current implementation is a cast and a single binop. The bounds implementation requires 2-4 binops and 2 reductions. Assuming binops are all reasonably comparable in terms of cost, I think we're basically asking if a cast is cheaper than two reductions which I could certainly benchmark for several scenarios.

Reply:
As another option I could translate what I wrote above into code and short circuit where we can, but fall back on the current approach if we can't.